juju/model.py

```diff
@@ -893,16 +918,15 @@ async def _all_watcher():
                         break
                     for delta in results.deltas:
                         try:
-                            delta = get_entity_delta(delta)
-                            old_obj, new_obj = self.state.apply_delta(delta)
-                            await self._notify_observers(delta, old_obj, new_obj)
-                        except KeyError as e:
-                            # TODO (stickupkid): we should raise the unknown delta
-                            # type, so we handle correctly all the types comming from
-                            # the all watcher. Currently they're ignored, causing
-                            # issue.
-                            # raise JujuError("unknown delta type {}".format(e.args))
-                            log.warning("unknown delta type: %s", e.args[0])
+                            entity = get_entity_delta(delta)
+                            old_obj, new_obj = self.state.apply_delta(entity)
+                            await self._notify_observers(entity, old_obj, new_obj)
+                            # Post step ensure that we can handle any settings
+                            # that need to be correctly set as a post step.
+                            _post_step(new_obj)
+                        except KeyError:
+                            if self.strict_mode:
+                                raise JujuError("unknown delta type '{}'".format(delta.entity))
```

Comment:
Which call does the KeyError come from?

An anti-pattern in Python is to take a low level error like KeyError thrown by a called function, and interpret it in the calling function. There might be other causes for KeyError, and it would be better to catch the error closer to the source, where we're more certain of the cause.

(This isn't a hill that I'm going to die on -- this might be a place where the pattern is not risky -- it just jumped out to me as I was reading the code.)

Reply:
I didn't write this originally, so I'm more than happy to try and fix this.